msal/src/Microsoft.Identity.Client/Exceptions/MsalExceptionFactory.cs

```diff
@@ -0,0 +1,158 @@
+            return GetServiceException(errorCode, errorMessage, null, null);
+        }
+
+        /// <summary>
+        /// Throws an <see cref="MsalClientException"/> exception
+        /// </summary>
+        /// <param name="errorCode">The error code</param>
+        /// <param name="errorMessage">A user friendly message</param>
+        /// <param name="exceptionDetail">More exception params</param>
+        public override Exception GetServiceException(
+            string errorCode, 
+            string errorMessage,
+            ExceptionDetail exceptionDetail = null)
+        {
+            ValidateRequiredArgs(errorCode, errorMessage);
+            return GetServiceException(errorCode, errorMessage, null, exceptionDetail);
+        }
+
+        /// <summary>
+        /// Throw an <see cref="MsalServiceException"/> exception
+        /// </summary>
+        public override Exception GetServiceException(
+            string errorCode, 
+            string errorMessage, 
+            Exception innerException = null, 
+            ExceptionDetail exceptionDetail = null)
+        {
+            ValidateRequiredArgs(errorCode, errorMessage);
+
+            return new MsalServiceException(
+                 errorCode,
+                 errorMessage,
+                 innerException)
+            {
+
+                Claims = exceptionDetail?.Claims,
+                ResponseBody = exceptionDetail?.ResponseBody,
+                StatusCode = exceptionDetail != null ? exceptionDetail.StatusCode : 0
+            };
+        }
+
+        /// <summary>
+        /// Throw an <see cref="MsalUiRequiredException"/>
+        /// </summary>
+        public override Exception GetUiRequiredException(
+            string errorCode, 
+            string errorMessage, 
+            Exception innerException,
+            ExceptionDetail exceptionDetail = null)
+        {
+            ValidateRequiredArgs(errorCode, errorMessage);
+
+            return new MsalUiRequiredException(errorCode, errorMessage, innerException)
+            {
+                Claims = exceptionDetail?.Claims,
+                ResponseBody = exceptionDetail?.ResponseBody,
+                StatusCode = exceptionDetail != null ? exceptionDetail.StatusCode : 0
+            };
+        }
+
+        public override string GetPiiScrubbedDetails(Exception ex)
+        {
+            var sb = new StringBuilder();
+            if (ex != null)
+            {
+                sb.AppendLine(String.Format(CultureInfo.CurrentCulture, "Exception type: {0}", ex.GetType()));
+
+                if (ex is MsalException)
+                {
+                    MsalException msalException = ex as MsalException;
+                    sb.AppendLine(String.Format(
+                        CultureInfo.CurrentCulture,
+                        ", ErrorCode: {0}",
+                        msalException.ErrorCode));
+                }
+
+                if (ex is MsalServiceException)
```

Comment:
if is...Exception could possibly also be moved to the factory class (allowing it to return a bool for these 2 cases). then we could perhaps move scrubbing to core? not a big deal and perhaps not worth the effort though.